MBS-10921: Drop wrongly listed edits from history

yvanzo opened the pull request:
# Problem

MBS-10921

# Solution

Drop all of these wrongly listed edits from history, at a limited hourly rate so as to not create further troubles with replication.

The whole process will take around 4 days.

# Action

1. Build a production image and deploy to run hourly cron.
2. Check progress during the week and update ticket’s status accordingly.
3. Remove the delete queries from the hourly script.

mwiencek commented on admin/cron/hourly.sh:
```diff
@@ -24,6 +24,81 @@ fi
+	JOIN recording r ON jsonb_extract_path_text(ed.data, 'entity0', 'id')::int = r.id
+	WHERE (e.type >= 90 AND e.type <= 99)
+	AND (jsonb_extract_path_text(ed.data, 'type0') = 'recording' AND jsonb_extract_path_text(ed.data, 'type1') = 'work')
+	AND r.id NOT IN (SELECT recording FROM recording_release rr WHERE er.release = rr.release)
```
PG seems to be having trouble optimizing the `recording_release` CTE, even though there's only a single use of it. This is much faster and *might* be equivalent:
```suggestion
      AND er.release NOT IN (SELECT release FROM medium m JOIN track t ON t.medium = m.id WHERE t.recording = r.id)
```

yvanzo replied:
Reworked in commit https://github.com/metabrainz/musicbrainz-server/pull/1581/commits/353ab4192d85de05b3cb9228b63a84e8abe13117 to get rid of CTEs.

yvanzo commented on admin/cron/hourly.sh:
```diff
@@ -24,6 +24,81 @@ fi
+	JOIN recording r ON jsonb_extract_path_text(ed.data, 'entity0', 'id')::int = r.id
+	WHERE (e.type >= 90 AND e.type <= 99)
+	AND (jsonb_extract_path_text(ed.data, 'type0') = 'recording' AND jsonb_extract_path_text(ed.data, 'type1') = 'work')
+	AND ea.artist NOT IN (SELECT artist FROM recording_artist ra WHERE r.id = ra.recording)
```
Fixed in commit https://github.com/metabrainz/musicbrainz-server/pull/1581/commits/353ab4192d85de05b3cb9228b63a84e8abe13117 by splitting each query in 4 queries (that makes a total of 12 queries) and by adjusting keys accordingly (but I didn’t find any `directly_related_entities`, any example?).

yvanzo commented on admin/cron/hourly.sh:
```diff
@@ -24,6 +24,81 @@ fi
+echo "MBS-10921 (1/3): Drop wrongly-listed edits in artist history..."
+
+./admin/psql READWRITE <<EOSQL
+SET LOCAL statement_timeout = 0;
```
Fixed in commit https://github.com/metabrainz/musicbrainz-server/pull/1581/commits/353ab4192d85de05b3cb9228b63a84e8abe13117 and set statement timeout to 15 minutes only.

mwiencek commented on admin/sql/updates/20200630-mbs-10921-unlist-unrelated-edits.sql:
```diff
@@ -0,0 +1,286 @@
+    AND e.type = 99
+    AND jsonb_extract_path_text(ed.data, 'link_type', 'entity0_type') = 'recording'
+    AND jsonb_extract_path_text(ed.data, 'link_type', 'entity1_type') = 'work'
+    AND jsonb_extract_path_text(data, 'relationship_order', '0', 'relationship', 'entity1', 'id')::int IN (
+        SELECT id FROM recording
+    )
+    AND jsonb_extract_path_text(data, 'relationship_order', '0', 'relationship', 'entity1', 'id')::int NOT IN (
+        SELECT r.id
+        FROM recording r
+        JOIN artist_credit ac ON r.artist_credit = ac.id
+        JOIN artist_credit_name acn ON ac.id = acn.artist_credit
+        WHERE acn.artist = ea.artist
+    )
+    LIMIT 50 -- of 534
```
That gets the recording (actually the work) at index 0, but `relationship_order` is a list of relationships:
https://github.com/metabrainz/musicbrainz-server/blob/4d3b7779ffd1d6753cd784fce21cd6f5b0cd3e8b/lib/MusicBrainz/Server/Edit/Relationship/Reorder.pm#L158-L159

`entity1` here gets the work, not the recording since type1 = work.

```suggestion
    AND EXISTS (
        SELECT 1
          FROM recording r
          JOIN (SELECT jsonb_array_elements(ed.data->'relationship_order') AS item) x
            ON (x.item#>'{relationship,entity0,id}')::INT = r.id
    )
    AND ea.artist NOT IN (
        SELECT acn.artist
          FROM artist_credit_name acn
          JOIN recording r ON r.artist_credit = acn.artist_credit
          JOIN (SELECT jsonb_array_elements(ed.data->'relationship_order') AS item) x
            ON (x.item#>'{relationship,entity0,id}')::INT = r.id
    )
    LIMIT 50 -- of 534
```

yvanzo replied:
Since it is just reordering recording-work relationships, I assume that every relationship holds the same `entity0.id`.
So I’m not sure it is necessary to run a more complex query. I will check edit data to confirm that.

mwiencek replied:
Oh ok, your analysis is probably correct here

yvanzo replied:
Fixed `s/entity1/entity0/` in https://github.com/metabrainz/musicbrainz-server/pull/1581/commits/ea6b10dd7562bed837af8b73ae7a49705bc12337.

mwiencek commented on admin/sql/updates/20200630-mbs-10921-unlist-unrelated-edits.sql:
```diff
@@ -0,0 +1,286 @@
+    AND e.type = 91
+    AND jsonb_extract_path_text(ed.data, 'type0') = 'recording'
+    AND jsonb_extract_path_text(ed.data, 'type1') = 'work'
+    AND jsonb_extract_path_text(ed.data, 'link', 'entity0', 'id')::int IN (
+        SELECT id FROM recording
+    )
+    AND jsonb_extract_path_text(ed.data, 'link', 'entity0', 'id')::int NOT IN (
+        SELECT r.id
+        FROM recording r
+        JOIN artist_credit ac ON r.artist_credit = ac.id
+        JOIN artist_credit_name acn ON ac.id = acn.artist_credit
+        WHERE acn.artist = ea.artist
+    )
```
https://github.com/metabrainz/musicbrainz-server/blob/4d3b7779ffd1d6753cd784fce21cd6f5b0cd3e8b/lib/MusicBrainz/Server/Edit/Relationship/Edit.pm#L267-L269 also output `$old->{entity0}` and `$new->{entity0}` (which would be present if the recording changed). Though `$old->{entity0}` should always be equal to `$link->{entity0}` if present. Maybe something like this?
```suggestion
    AND EXISTS (
        SELECT 1
          FROM recording
         WHERE id IN (VALUES
             (jsonb_extract_path_text(ed.data, 'link', 'entity0', 'id')::int),
             (jsonb_extract_path_text(ed.data, 'new', 'entity0', 'id')::int))
    )
    AND ea.artist NOT IN (
        SELECT acn.artist
          FROM artist_credit_name acn
          JOIN recording r ON r.artist_credit = acn.artist_credit
         WHERE r.id IN (VALUES
             (jsonb_extract_path_text(ed.data, 'link', 'entity0', 'id')::int),
             (jsonb_extract_path_text(ed.data, 'new', 'entity0', 'id')::int))
    )
```

yvanzo replied:
Picked up as https://github.com/metabrainz/musicbrainz-server/pull/1581/commits/9eb0f6b311933a8e7c003020c1ad4a14a537d6f8, amended with https://github.com/metabrainz/musicbrainz-server/pull/1581/commits/46ff2a3265a23d7e9b4339297a5151812576d39f.

yvanzo replied:
That is actually unneeded, reverted in https://github.com/metabrainz/musicbrainz-server/pull/1581/commits/29074454c43a5123ee86f3a71d75594f5ac0fdd5.

mwiencek commented on admin/sql/updates/20200630-mbs-10921-unlist-unrelated-edits.sql:
```diff
@@ -0,0 +1,286 @@
+    AND jsonb_extract_path_text(ed.data, 'type0') = 'recording'
+    AND jsonb_extract_path_text(ed.data, 'type1') = 'work'
+    AND jsonb_extract_path_text(ed.data, 'entity0', 'id')::int != er.recording
+    AND jsonb_extract_path_text(ed.data, 'entity0', 'id')::int IN (
+        SELECT id FROM recording
+    )
```
I'm not sure it matters that the linked recording still exists in this case, since if the ID is different we know it's wrong in either case?

yvanzo replied:
It doesn’t seem to be necessarily wrong. I supposed it might also be the recording id has just changed after a merge. I didn’t check if that is actually possible yet.

mwiencek replied:
Yeah, merged recordings won't be handled by any of the queries at present. You'd have to find the new ID via the `recording_gid_redirect` table. Of course I'm not sure that's worth the added effort or slowdown.

mwiencek commented on admin/sql/updates/20200630-mbs-10921-unlist-unrelated-edits.sql:
```diff
@@ -0,0 +1,286 @@
+    AND e.type = 91
+    AND jsonb_extract_path_text(ed.data, 'type0') = 'recording'
+    AND jsonb_extract_path_text(ed.data, 'type1') = 'work'
+    AND jsonb_extract_path_text(ed.data, 'link', 'entity0', 'id')::int != er.recording
+    AND jsonb_extract_path_text(ed.data, 'link', 'entity0', 'id')::int IN (
+        SELECT id FROM recording
+    )
```
Like for artists above, this should handle `{new,entity0,id}` too.

yvanzo replied:
Added in https://github.com/metabrainz/musicbrainz-server/pull/1581/commits/154df69f97b335db2c8d4cd499685129eeeebea0

yvanzo replied:
Actually unneeded, reverted in https://github.com/metabrainz/musicbrainz-server/pull/1581/commits/29074454c43a5123ee86f3a71d75594f5ac0fdd5.

mwiencek commented on admin/sql/updates/20200630-mbs-10921-unlist-unrelated-edits.sql:
```diff
@@ -0,0 +1,286 @@
+    AND e.type = 99
+    AND jsonb_extract_path_text(ed.data, 'link_type', 'entity0_type') = 'recording'
+    AND jsonb_extract_path_text(ed.data, 'link_type', 'entity1_type') = 'work'
+    AND jsonb_extract_path_text(data, 'relationship_order', '0', 'relationship', 'entity1', 'id')::int != er.recording
+    AND jsonb_extract_path_text(data, 'relationship_order', '0', 'relationship', 'entity1', 'id')::int IN (
+        SELECT id FROM recording
+    )
```
Same comment as for reorder relationships in edit_artist above.

yvanzo replied:
Unneeded, same as https://github.com/metabrainz/musicbrainz-server/pull/1581#discussion_r448609981.

(Fixed `s/entity1/entity0/` in https://github.com/metabrainz/musicbrainz-server/pull/1581/commits/ea6b10dd7562bed837af8b73ae7a49705bc12337 as reported in https://github.com/metabrainz/musicbrainz-server/pull/1581#discussion_r448567539.)

mwiencek commented on admin/sql/updates/20200630-mbs-10921-unlist-unrelated-edits.sql:
```diff
@@ -0,0 +1,286 @@
+    AND e.type = 91
+    AND jsonb_extract_path_text(ed.data, 'type0') = 'recording'
+    AND jsonb_extract_path_text(ed.data, 'type1') = 'work'
+    AND jsonb_extract_path_text(ed.data, 'link', 'entity0', 'id')::int IN (
+        SELECT id FROM recording
+    )
+    AND jsonb_extract_path_text(ed.data, 'link', 'entity0', 'id')::int NOT IN (
+        SELECT track.recording
+        FROM track
+        JOIN medium ON track.medium = medium.id
+        WHERE medium.release = er.release
+    )
```
Same here re `{new,entity0,id}`.

yvanzo replied:
Added in https://github.com/metabrainz/musicbrainz-server/pull/1581/commits/04969b79a9ea8489783c7bd7090e73578593917e

yvanzo replied:
That is actually not needed either, see https://github.com/metabrainz/musicbrainz-server/pull/1581/commits/29074454c43a5123ee86f3a71d75594f5ac0fdd5.

mwiencek commented on admin/sql/updates/20200630-mbs-10921-unlist-unrelated-edits.sql:
```diff
@@ -0,0 +1,286 @@
+    AND e.type = 99
+    AND jsonb_extract_path_text(ed.data, 'link_type', 'entity0_type') = 'recording'
+    AND jsonb_extract_path_text(ed.data, 'link_type', 'entity1_type') = 'work'
+    AND jsonb_extract_path_text(data, 'relationship_order', '0', 'relationship', 'entity1', 'id')::int IN (
+        SELECT id FROM recording
+    )
+    AND jsonb_extract_path_text(data, 'relationship_order', '0', 'relationship', 'entity1', 'id')::int NOT IN (
+        SELECT track.recording
+        FROM track
+        JOIN medium ON track.medium = medium.id
+        WHERE medium.release = er.release
+    )
```
Same here re handling every relationship_order item and s/entity1/entity0/.

yvanzo replied:
Thanks!

yvanzo replied:
> Same here re handling every relationship_order item

Unneeded, same as https://github.com/metabrainz/musicbrainz-server/pull/1581#discussion_r448609981.

> and s/entity1/entity0/.

Fixed in https://github.com/metabrainz/musicbrainz-server/pull/1581/commits/ea6b10dd7562bed837af8b73ae7a49705bc12337.

yvanzo commented on admin/sql/updates/20200630-mbs-10921-unlist-unrelated-edits.sql:
```diff
@@ -0,0 +1,286 @@
+    FROM edit_artist ea
+    JOIN edit e ON ea.edit = e.id
+    JOIN edit_data ed ON e.id = ed.edit
+    WHERE e.id >= 70658947 AND e.id < 71006471
```
Yes, see the commit message.

yvanzo commented on admin/sql/updates/20200630-mbs-10921-unlist-unrelated-edits.sql:
```diff
@@ -0,0 +1,286 @@
+        SELECT r.id
+        FROM recording r
+        JOIN artist_credit ac ON r.artist_credit = ac.id
+        JOIN artist_credit_name acn ON ac.id = acn.artist_credit
```
Picked it up

yvanzo replied:
Applied the same change to other queries in https://github.com/metabrainz/musicbrainz-server/pull/1581/commits/082dcea278363bec428ae3e099eb642f982fb32d.

yvanzo commented:
Fixed everything back and forth, and tested it again.

About 230K edits won’t be removed due to the last 3 months of recordings merge, see https://github.com/metabrainz/musicbrainz-server/pull/1581/commits/577180736c8faa59b5cdafea6fd571a118aece49.

yvanzo commented:
Rebased and squashed patches into one commit https://github.com/metabrainz/musicbrainz-server/pull/1581/commits/2c5136a89d0ec3830f290bd327c66b6c4f8dbe0a.

mwiencek commented on admin/sql/updates/20200921-mbs-10921-unlist-unrelated-edits.sql:
```diff
@@ -0,0 +1,282 @@
+    AND jsonb_extract_path_text(ed.data, 'relationship', 'entity0', 'id')::int NOT IN (
+        SELECT r.id
+        FROM recording r
+        JOIN artist_credit_name acn ON recording.artist_credit = acn.artist_credit
```
```suggestion
        JOIN artist_credit_name acn ON r.artist_credit = acn.artist_credit
```

yvanzo replied:
Picked it up with https://github.com/metabrainz/musicbrainz-server/pull/1581/commits/44ccba42ff3e8804c7a8f66f3619aaee393ea2ff.

mwiencek commented on admin/sql/updates/20200921-mbs-10921-unlist-unrelated-edits.sql:
```diff
@@ -0,0 +1,282 @@
+    AND jsonb_extract_path_text(data, 'relationship_order', '0', 'relationship', 'entity0', 'id')::int NOT IN (
+        SELECT r.id
+        FROM recording r
+        JOIN artist_credit_name acn ON recording.artist_credit = acn.artist_credit
```
```suggestion
        JOIN artist_credit_name acn ON r.artist_credit = acn.artist_credit
```

yvanzo replied:
Picked it up with https://github.com/metabrainz/musicbrainz-server/pull/1581/commits/44ccba42ff3e8804c7a8f66f3619aaee393ea2ff.

mwiencek commented on admin/sql/updates/20200921-mbs-10921-unlist-unrelated-edits.sql:
```diff
@@ -0,0 +1,282 @@
+    FROM edit_recording er
+    JOIN edit e ON er.edit = e.id
+    JOIN edit_data ed ON er.edit = ed.edit
+    WHERE er.edit >= 70658947 AND er.edit < 71006471
```
using e.id allows PG to optimize things much better here
```suggestion
    WHERE e.id >= 70658947 AND e.id < 71006471
```

mwiencek commented on admin/sql/updates/20200921-mbs-10921-unlist-unrelated-edits.sql:
```diff
@@ -0,0 +1,282 @@
+    FROM edit_recording er
+    JOIN edit e ON er.edit = e.id
+    JOIN edit_data ed ON er.edit = ed.edit
+    WHERE er.edit >= 70658947 AND er.edit < 71006471
```
```suggestion
    WHERE e.id >= 70658947 AND e.id < 71006471
```

mwiencek commented on admin/sql/updates/20200921-mbs-10921-unlist-unrelated-edits.sql:
```diff
@@ -0,0 +1,282 @@
+    FROM edit_recording er
+    JOIN edit e ON er.edit = e.id
+    JOIN edit_data ed ON er.edit = ed.edit
+    WHERE er.edit >= 70658947 AND er.edit < 71006471
```
```suggestion
    WHERE e.id >= 70658947 AND e.id < 71006471
```

mwiencek commented on admin/sql/updates/20200921-mbs-10921-unlist-unrelated-edits.sql:
```diff
@@ -0,0 +1,282 @@
+    FROM edit_recording er
+    JOIN edit e ON er.edit = e.id
+    JOIN edit_data ed ON er.edit = ed.edit
+    WHERE er.edit >= 70658947 AND er.edit < 71006471
```
```suggestion
    WHERE e.id >= 70658947 AND e.id < 71006471
```

yvanzo commented on admin/sql/updates/20200921-mbs-10921-unlist-unrelated-edits.sql:
```diff
@@ -0,0 +1,282 @@
+    AND jsonb_extract_path_text(ed.data, 'link', 'entity0', 'id')::int NOT IN (
+        SELECT r.id
+        FROM recording r
+        JOIN artist_credit_name acn ON recording.artist_credit = acn.artist_credit
```
Picked it up with https://github.com/metabrainz/musicbrainz-server/pull/1581/commits/44ccba42ff3e8804c7a8f66f3619aaee393ea2ff.
